File: src/Spreads.Collections/MutableSeries.fs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.


namespace Spreads

open System
open System.Linq
open System.Linq.Expressions
open System.Collections.Generic
open System.Diagnostics
open System.Runtime.InteropServices
open System.Runtime.CompilerServices
open System.Threading
open System.Threading.Tasks
open System.Numerics

open Spreads
open Spreads.Collections

[<AllowNullLiteral>]
[<AbstractClassAttribute>]
//[<DebuggerTypeProxy(typeof<SeriesDebuggerProxy<_,_>>)>]
type MutableSeries<'K,'V> internal() as this =
  inherit Series<'K,'V>()

  abstract member Add : key:'TKey * value:'TValue -> unit
  abstract member AddFirst : key:'TKey * value:'TValue -> unit
  abstract member AddLast : key:'TKey * value:'TValue -> unit
  abstract member Append : appendMap:IReadOnlySeries<'TKey,'TValue> * option:AppendOption -> int
  abstract member Complete : unit -> unit
  abstract member Count : int64
  abstract member Item : 'TValue with get, set
  abstract member Remove : key:'TKey -> bool
  abstract member RemoveFirst : kvp:byref<KeyValuePair<'TKey,'TValue>> -> bool
  abstract member RemoveLast : kvp:byref<KeyValuePair<'TKey,'TValue>> -> bool
  abstract member RemoveMany : key:'TKey * direction:Lookup -> bool
  abstract member Version : int64 with get

  interface IReadOnlySeries<'K,'V> with
    member x.Comparer = this.Comparer
    member x.First = this.First
    member x.GetAt(idx) = this.GetAt(idx)
    member x.GetCursor() = this.GetCursor()
    member x.GetEnumerator(): Collections.IEnumerator = this.GetCursor() :> Collections.IEnumerator
    member x.GetEnumerator(): IAsyncEnumerator<KeyValuePair<'K,'V>> = this.GetCursor() :> IAsyncEnumerator<KeyValuePair<'K,'V>>
    member x.GetEnumerator(): IEnumerator<KeyValuePair<'K,'V>> = this.GetCursor() :> IEnumerator<KeyValuePair<'K,'V>>
    member x.IsEmpty = this.IsEmpty
    member x.IsIndexed = this.IsIndexed
    member x.IsReadOnly = this.IsReadOnly
    member x.Item with get (key) = this.Item(key)
    member x.Keys = this.Keys
    member x.Last = this.Last
    member x.Subscribe(observer) = this.Subscribe(observer)
    member x.SyncRoot = this.SyncRoot
    member this.TryFind(k:'K, direction:Lookup, [<Out>] result: byref<KeyValuePair<'K, 'V>>) = this.TryFind(k, direction, &result)
    member this.TryGetFirst([<Out>] res: byref<KeyValuePair<'K, 'V>>) = this.TryGetFirst(&res)
    member this.TryGetLast([<Out>] res: byref<KeyValuePair<'K, 'V>>) = this.TryGetLast(&res)
    member this.TryGetValue(k, [<Out>] value:byref<'V>) = this.TryGetValue(k, &value)
    member x.Values = this.Values

  interface IMutableSeries<'K,'V> with
    member this.Complete() = this.Complete()
    member this.Version with get() = this.Version
    member this.Count with get() = this.Count
    member this.Item with get k = this.Item(k) and set (k:'K) (v:'V) = this.Item(k,v)
    member this.Add(k, v) = this.Add(k,v)
    member this.AddLast(k, v) = this.AddLast(k, v)
    member this.AddFirst(k, v) = this.AddFirst(k, v)
    member this.Append(appendMap:IReadOnlySeries<'K,'V>, option:AppendOption) = this.Append(appendMap, option)
    member this.Remove(k) = this.Remove(k)
    member this.RemoveFirst([<Out>] result: byref<KeyValuePair<'K, 'V>>) = this.RemoveFirst(&result)
    member this.RemoveLast([<Out>] result: byref<KeyValuePair<'K, 'V>>) = this.RemoveLast(&result)
    member this.RemoveMany(key:'K,direction:Lookup) = this.RemoveMany(key, direction)


[<AllowNullLiteral>]
//[<DebuggerTypeProxy(typeof<SeriesDebuggerProxy<_,_>>)>]
type internal MutableSeriesWrapper<'K,'V> internal(inner:IMutableSeries<'K,'V>) as this =
  inherit MutableSeries<'K,'V>()
  let foo = "a"

File: src/Spreads/Series/MutableSeries.cs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

using System;
using System.Collections.Generic;

namespace Spreads.Series
{
    [Obsolete("Use interface")]
    public class MutableSeries<TK, TV> : Series<TK, TV>, IMutableSeries<TK, TV>
    {
        private readonly IMutableSeries<TK, TV> _innerSeries;

        public MutableSeries(IMutableSeries<TK, TV> innerSeries)
        {
            _innerSeries = innerSeries;
        }

        public long Count => _innerSeries.Count;

        public long Version => _innerSeries.Version;

        public override bool IsReadOnly => _innerSeries.IsReadOnly;

        public override KeyComparer<TK> Comparer => _innerSeries.Comparer;

        public override bool IsIndexed => _innerSeries.IsIndexed;

        public override bool IsEmpty => _innerSeries.IsEmpty;

        public override KeyValuePair<TK, TV> First => _innerSeries.First;

        public override KeyValuePair<TK, TV> Last => _innerSeries.Last;

        TV IMutableSeries<TK, TV>.this[TK key]
        {
            get { return _innerSeries[key]; }
            set { _innerSeries[key] = value; }
        }

        public override IEnumerable<TK> Keys => _innerSeries.Keys;

        public override IEnumerable<TV> Values => _innerSeries.Values;

        public override bool TryFind(TK key, Lookup direction, out KeyValuePair<TK, TV> value)
        {
            return _innerSeries.TryFind(key, direction, out value);
        }

        public override bool TryGetFirst(out KeyValuePair<TK, TV> value)
        {
            return _innerSeries.TryGetFirst(out value);
        }

        public override bool TryGetLast(out KeyValuePair<TK, TV> value)
        {
            return _innerSeries.TryGetLast(out value);
        }

        public void Add(TK key, TV value)
        {
            _innerSeries.Add(key, value);
        }

        public void AddLast(TK key, TV value)
        {
            _innerSeries.AddLast(key, value);
        }

        public void AddFirst(TK key, TV value)
        {
            _innerSeries.AddFirst(key, value);
        }

        public bool Remove(TK key)
        {
            return _innerSeries.Remove(key);
        }

        public bool RemoveLast(out KeyValuePair<TK, TV> kvp)
        {
            return _innerSeries.RemoveLast(out kvp);
        }

        public bool RemoveFirst(out KeyValuePair<TK, TV> kvp)
        {
            return _innerSeries.RemoveFirst(out kvp);
        }

        public bool RemoveMany(TK key, Lookup direction)
        {
            return _innerSeries.RemoveMany(key, direction);
        }

        public int Append(IReadOnlySeries<TK, TV> appendMap, AppendOption option)
        {
            return _innerSeries.Append(appendMap, option);
        }

        public void Complete()
        {
            _innerSeries.Complete();
        }

        public override ICursor<TK, TV> GetCursor()
        {
            throw new NotImplementedException();
        }
    }
}